File: 2020/day6.py
from functools import reduce

with open("day6.txt") as f:
    data = f.read()


groups = data.split("\n\n")

gd = {}

for ix, group in enumerate(groups):
    answers = []
    for line in group.split("\n"):
        answers.append(set(line))
    gd[ix] = reduce(lambda x, y: x | y, answers)

print(sum(map(lambda x: len(x), gd.values())))

for ix, group in enumerate(groups):
    answers = []
    for line in group.split("\n"):
        answers.append(set(line))
    gd[ix] = reduce(lambda x, y: x & y, answers)

print(sum(map(lambda x: len(x), gd.values())))

File: 2020/day6.exs
defmodule Day6 do
  def run do
    File.read!("day6.txt")
    |> String.split("\n\n")
    |> Enum.reduce({0, 0}, &count/2)
  end

  def count(lines, {all, some}) do
    [union, intersection] =
      String.split(lines, "\n")
      |> Enum.map(&to_set/1)
      |> Enum.reduce({0, 0}, &reduce/2)
      |> Enum.map(&MapSet.size/1)

    {all + union, intersection + some}
  end

  def to_set(string) do
    string
    |> String.graphemes()
    |> MapSet.new()
  end

  def reduce(string_set, {0, 0}) do
    [string_set, string_set]
  end

  def reduce(string_set, [all, some]) do
    [MapSet.union(string_set, all), MapSet.intersection(string_set, some)]
  end
end
